test_can_restrict_endianness fails on big-endian s390x

Hi!

**test_can_restrict_endianness** fails on s390x, the only big-endian architecture being tested in Ubuntu.
Links to full tests logs can be found on the following page:
http://autopkgtest.ubuntu.com/packages/p/python-hypothesis/bionic/s390x

Is this by design?

The test fails when **dt=dtype('int16')** and **dt.byteorder == '='**
```
=================================== FAILURES ===================================
_________________________ test_can_restrict_endianness _________________________

    @given(nps.integer_dtypes(endianness='>'))
>   def test_can_restrict_endianness(dt):

tests/numpy/test_gen_data.py:189: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib/python3/dist-packages/hypothesis/core.py:579: in execute
    result = self.test_runner(data, run)
/usr/lib/python3/dist-packages/hypothesis/executors.py:58: in default_new_style_executor
    return function(data)
/usr/lib/python3/dist-packages/hypothesis/core.py:571: in run
    return test(*args, **kwargs)
tests/numpy/test_gen_data.py:189: in test_can_restrict_endianness
    def test_can_restrict_endianness(dt):
/usr/lib/python3/dist-packages/hypothesis/core.py:518: in test
    result = self.test(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

dt = dtype('int16')

    @given(nps.integer_dtypes(endianness='>'))
    def test_can_restrict_endianness(dt):
        if dt.itemsize == 1:
            assert dt.byteorder == '|'
        else:
>           assert dt.byteorder == '>'
E           AssertionError

tests/numpy/test_gen_data.py:193: AssertionError
---------------------------------- Hypothesis ----------------------------------
Falsifying example: test_can_restrict_endianness(dt=dtype('int16'))
======== 1 failed, 2156 passed, 4 skipped, 1 xfailed in 278.16 seconds =========
```

--- a/RELEASE.rst
+++ b/RELEASE.rst
@@ -1,0 +1,8 @@
+RELEASE_TYPE: patch
+
+This patch fixes our tests for Numpy dtype strategies on big-endian platforms,
+where the strategy behaved correctly but the test assumed that the native byte
+order was little-endian.
+
+There is no user impact unless you are running our test suite on big-endian
+platforms.  Thanks to Graham Inggs for reporting :issue:`1164`.

--- a/tests/numpy/test_gen_data.py
+++ b/tests/numpy/test_gen_data.py
@@ -17,6 +17,8 @@
 
 from __future__ import division, print_function, absolute_import
 
+import sys
+
 import numpy as np
 import pytest
 from flaky import flaky
@@ -185,12 +187,14 @@ def test_can_turn_off_subarrays(dt):
         assert field.shape == ()
 
 
-@given(nps.integer_dtypes(endianness='>'))
-def test_can_restrict_endianness(dt):
-    if dt.itemsize == 1:
-        assert dt.byteorder == '|'
+@pytest.mark.parametrize('byteorder', ['<', '>'])
+@given(data=st.data())
+def test_can_restrict_endianness(data, byteorder):
+    dtype = data.draw(nps.integer_dtypes(byteorder, sizes=(16, 32, 64)))
+    if byteorder == ('<' if sys.byteorder == 'little' else '>'):
+        assert dtype.byteorder == '='
     else:
-        assert dt.byteorder == '>'
+        assert dtype.byteorder == byteorder
 
 
 @given(nps.integer_dtypes(sizes=8))

